x/dogfood/keeper/abci.go

```diff
@@ -1,12 +1,21 @@
 package keeper
 
 import (
-	"github.com/ExocoreNetwork/exocore/x/dogfood/types"
+	"sort"
+
 	abci "github.com/cometbft/cometbft/abci/types"
+	tmprotocrypto "github.com/cometbft/cometbft/proto/tendermint/crypto"
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
 func (k Keeper) EndBlock(ctx sdk.Context) []abci.ValidatorUpdate {
+	id, _ := k.getValidatorSetID(ctx, ctx.BlockHeight())
+	if !k.IsEpochEnd(ctx) {
+		// save the same id for the next block height.
+		k.setValidatorSetID(ctx, ctx.BlockHeight()+1, id)
+		return []abci.ValidatorUpdate{}
+	}
+	defer k.ClearEpochEnd(ctx)
 	// start with clearing the hold on the undelegations.
```

Comment:
I haven't seen the code to increase the undelegationHoldCount. Why does it need to be decreased here? Additionally, I have looked at the related code about OnHold in Cosmos-SDK and Evmos, It's a great feature. But I didn't find this feature being used in Cosmos-SDK and Evmos.

Reply:
> I haven't seen the code to increase the undelegationHoldCount. Why does it need to be decreased here? 

You are right; I made a mistake. The line that increases it was accidentally removed. I will add it back. Great catch!

> Additionally, I have looked at the related code about OnHold in Cosmos-SDK and Evmos, It's a great feature. But I didn't find this feature being used in Cosmos-SDK and Evmos.

Evmos certainly does not use it, as far as I can tell. Cosmos-SDK introduced the feature to support interchain-security, which does use it. I do not think any standalone chains use it.